warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java

```diff
@@ -150,18 +144,21 @@ public Value reduce(Key key, Iterator<Value> iter) {
                 long delta = v.getCOUNT();
                 
                 count += delta;
-                /**
+                /*
                  * Fail fast approach.
                  */
                 if (v.getIGNORE()) {
                     seenIgnore = true;
-                    if (log.isDebugEnabled())
-                        log.debug("SeenIgnore is true. Skipping collections");
+                    log.debug("SeenIgnore is true. Skipping collections");
                 }
                 
-                // if delta > 0, we are collecting the uid list
-                // in the protobuf into our object's uid list.
-                if (delta > 0) {
+                // if delta >= 0, we are collecting the uid list in the protobuf into our object's uid list.
+                //
+                // Note that it is possible to have a delta of 0 if we were to have gone through
+                // a partial major compaction and combined a protocol buffer having a positive
+                // count of 1 and a negative count of 1 (leaving one UID in the UID list and one
+                // UID in the REMOVEDUID list.
+                if (delta >= 0) {
                     
                     for (String uid : v.getQUARANTINEUIDList()) {
```

Comment:
I'm not fulling understanding the construction of this protobuf. For the case where delta is non-negative, we appear to be treating the QUARANTINEUIDList as "released UIDs" which we'll add back into the list of uids in the aggregator call. For negative delta, this is treated as a quarantine list not to be included in the uids. Is this a correct understanding?

Reply:
Yeah, I'm not really sure about the quarantined/released lists. Maybe @phrocker can comment? Think he knows...

Reply:
It does look like maybe the intent was for the QUARANTINEDUID list to be dual-purpose... You quarantine a UID by doing an insert with a count of -1 and the UID in the QUARANTINEDUID list. You release it by doing an insert with a count of 1 and the UID in the QUARANTINEDUID list. However, with partial major compactions, that approach won't work (or I'm just wrong about how the feature works) since the -1 count PB with the uid in the QUARANTINEDUID list could get combined with a PB having, say, a count of 2 and two other uids in the UID list. The result would be a PB with a count of 1, 2 uids in the UID list, and 1 in the QUARANTINEDUID list. The next time that PB is read in (for any of a scan, partial major compaction, or full major compaction) the delta will be positive and the quarantined UID will be released even though that probably wasn't the intent.

Reply:
hmm, my recollection is that all keys would be propagated until a full major C, so that even if the partial major occurred all keys would exist until that FullMajorC was launched periodically.

Reply:
In the untouched version of this, the uids in the QUARANTINEDUID list are released when the delta is positive (see [here](https://github.com/NationalSecurityAgency/datawave/blob/master/warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java#L166)) and uids in the QUARANTINEDUID list are carried forward as quarantined when the delta is negative (see [here](https://github.com/NationalSecurityAgency/datawave/blob/master/warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java#L204)).

This seems like a problem for partial major compactions. In case 1, say two PBs are combined. One has a count of -1 and uid1 in the QUARANTINEDUID list. The other PB has a count of 2 and uid2 and uid3 in the UID list. The combined PB will have a count of 1, uid2 and uid3 in the UID list, and uid1 in the QUARANTINEDUID list. Then, next time there's a scan or another compaction, the incoming delta for this new PB will be positive and the quarantined uid1 will be released. For case 2, say a previously quarantined uid1 is released so there's a PB with a count of 1 and uid1 in the QUARANTINEDUID list. Say that PB gets combined with another during a partial major compaction where the other PB has a count of -2 with uid2 and uid3 in the REMOVEDUID list. The final PB will have a count of -1, uid1 in the UID list, and uid2 and uid3 in the REMOVEDUID list. Now when that PB eventually gets combined with the one that originally quarantined uid1, uid1 will be re-quarantined.

Reply:
Yep. I thought there being checks for testing if  partial, not doing any combining of keys at all, leaving scan and full major c to do the work as a safeguard....then periodic full majors doing the combining and reduction of keys...but again, my knowledge is vestigial.

Reply:
Perhaps the way to handle this is to modify the protobuf to add a RELEASEDUID list as well. When the intent is to release a previously quarantined UID, insert a PB with a count of 1 and the uid in the new RELEASEDUID list. Then the action to take in the code is clear and it will work with partial major compactions. We could add a special case to release a UID if the incoming PB contains only uids in the QUARANTINEDUID list and not any of the other lists. That would catch previously released quarantined uids that were never involved in a partial major compaction. For those that were involved in a partial majc, I don't think one can make any assumptions about uids in the QUARANTINEDUID list. @ivakegg, any thoughts? I have some local changes to attempt this and also fix the case @jschmidt10  identified. I'll go ahead and push those, and we can just drop the new commit if people don't like that approach.

Reply:
@javease, I believe my latest change should address your question and also resolve the discussion in this thread.

Reply:
The QUARANTINEDUID list has been removed, so this discussion is OBE. Resolving...